### PR TITLE
feat: set value of `submit` fields when form is submitted

### PR DESCRIPTION
**Pull request by Rich-Harris**
alternative to #15970. could probably use a test

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

**Comment by lettucebowler** [comment]
Perhaps a strange use-case but the solution above keeps the data on the next submit if first a named submitter is used and then another without a name or a different name is used.

first submit: <button name="extraAction" value="example">extra example</button>

instance.fields.value(): { ..., extraAction: 'example' }

second submit with <button type="submit">no extra</button>

instance.fields.value(): { ..., extraAction: 'example' } // it's still there

Now you have to keep track of which values are present in form_data to know if they should still be in instance.fields.value() in case something like this happens.

**Comment by Rich-Harris** [comment]
Great catch! I _think_ we can solve it by just clearing whatever the previous `event.submitter` value was each time we submit.

**Comment by ottomated** [comment]
> Great catch! I _think_ we can solve it by just clearing whatever the previous `event.submitter` value was each time we submit.

What if there's a submitter with the same name as another field? Not a common scenario but it is valid html.

**Comment by lettucebowler** [comment]
> > Great catch! I _think_ we can solve it by just clearing whatever the previous `event.submitter` value was each time we submit.
> 
> What if there's a submitter with the same name as another field? Not a common scenario but it is valid html.

Its possible for multiple fields to have the same name with a standard form. The FormData would have multiple entries that you can get with .getAll() similar to how search parameters work. I think with the remote form that field would need to be configured as an array in the schema to get both values, otherwise it would probably fail at the validation step.

Handle_submit would need to also check for that and append it to the array if present, otherwise add the array to the data. Then that has to also be handled when clearing out the data if a different submitter is used on the next submit.

**Comment by Rich-Harris** [comment]
> Its possible for multiple fields to have the same name with a standard form

Yes, but not in a SvelteKit form. With `FormData`, you can have a field `foo` that is either a single value (`fd.get('foo')`) or an array (`fd.getAll('foo')`), but since there's no sensible way to convert it to a POJO (because you can't distinguish between single values and arrays with only one member), duplicates are banned.

So I'm not seeing a downside to this approach.

**Comment by ottomated** [comment]
This will break: if the user clicks "Add +1" then submits a custom increment, it will clear the `amount` field.

```svelte
<script>
  import { f } from './form.remote';
</script>

<form {...f}>
<input {...f.fields.amount.as('number')}>
<button type="submit">Custom increment</button>

<button {...f.fields.amount.as('submit', 1)}>Add +1</button>
<button {...f.fields.amount.as('submit', 5)}>Add +5</button>
</form>
```

**Comment by Rich-Harris** [comment]
Did you try that code? It's invalid — you can either have `<button>` elements using `amount.as('submit', ...)` _or_ an `<input>` element using `amount.as('number')`, but you can't have both

**Comment by ottomated** [comment]
You're right, here's a valid example then:
```svelte
<script>
	import { my_form } from './form.remote';
	
	let show_custom = $state(false);
</script>

<label>
	Custom amount:
	<input type="checkbox" bind:checked={show_custom} />
</label>
<form {...my_form}>
	{#if show_custom}
		<input {...my_form.fields.amount.as('number')}>
		<button type="submit">Custom increment</button>
	{:else}
		<button {...my_form.fields.amount.as('submit', 1)}>Add +1</button>
		<button {...my_form.fields.amount.as('submit', 5)}>Add +5</button>
	{/if}
</form>

<pre>{JSON.stringify(my_form.result, null, 2)}</pre>
```

**Comment by ottomated** [comment]
Another (potential) issue - have you considered two instances of the same form? Would clearing the submitter of one clear the submitter of the other?

**Comment by Rich-Harris** [comment]
> You're right, here's a valid example then:

What is this showing? I'm not sure what the difference between the expected/actual behaviour is

> have you considered two instances of the same form?

You can't _have_ two instances of the same form:

image

**Comment by ottomated** [comment]
> What is this showing? I'm not sure what the difference between the expected/actual behaviour is

1. Press +5, observe `form.fields.amount.value() == 5`
2. Check custom amount
3. Enter custom amount N and submit, observe `form.fields.amount.value() == undefined` instead of N because the previous submitter has been incorrectly cleared.

**Comment by ottomated** [comment]
> You can't _have_ two instances of the same form:

Sorry, bad phrasing on my part; the form instance is reused if the same form is on 2 different pages. Tested with this branch locally and it's pretty much the same problem as my other example so the same fix should work for both - I made a PR #16152 that fixes it

**Comment by Rich-Harris** [comment]
I think the fix is much simpler, no? 10110eddcdf69d8faae18c8bb12199710e8a5d9f

**Comment by ottomated** [comment]
> I think the fix is much simpler, no? [10110ed](https://github.com/sveltejs/kit/commit/10110eddcdf69d8faae18c8bb12199710e8a5d9f)

Submission 1 (`submitter == 'n:foo'`): `{ 'n:foo': 123 }`; `previous_submitter_name = 'n:foo'`

Submission 2 (no submitter, foo is a text field): `{ foo: 'str' }`
`form_data.has('n:foo') === false`, so `foo` would be incorrectly cleared.

**Comment by Rich-Harris** [comment]
fixed

**Comment by ottomated** [comment]
That's all I can see, sorry for nitpicking so much!

**Comment by Rich-Harris** [comment]
Reporting bugs is never nitpicking!

File: packages/kit/src/runtime/client/remote-functions/form.svelte.js [hunk 1]
```diff
@@ -402,6 +405,29 @@ export function form(id) {
...
+					// `set_nested_value` would coerce `undefined` to `NaN`/`false`
+					// instead of clearing the previously-submitted value.
+					set_nested_value(input, previous_submitter_name, undefined);
+				}
+
+				if (event.submitter) {
+					const name = event.submitter.getAttribute('name');
+					const value = /** @type {any} */ (event.submitter).value;
+
+					if (name !== null && value !== undefined) {
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/runtime/client/remote-functions/form.svelte.js:422]: the fact that this line is correct and has comparisons to both `null` and `undefined` is proof javascript is trash

**Review by elliott-with-the-longest-name-on-github — APPROVED** [review]

**Comment by elliott-with-the-longest-name-on-github** [comment]
One nitpick: Should you add some regression tests for the edge cases ottomated pointed out?

**Comment by ottomated** [comment]
You could take the tests from #16152
